### DISCUSSION TITLE
Move window to an specific Display

### DISCUSSION THREAD

**Discussion by carlos-sarmiento** [rxhanson/Rectangle#401] [Ideas]
I'm interested in a feature in which you could position a window on a specific display. For example if you press, Cmd + Ctrl + {number} it would be sent to that display.

I've never programmed on Swift, but I would be happy to work on the implementation if the maintainers are willing to accept it into the project.

**Comment by rxhanson** [comment]
Sounds good to me - I like the idea and I can see merging a PR with this functionality. At first cut, I think I'd prefer it as a Terminal config. 

A couple things to keep in mind on it, off the top of my head.

- The screens macOS API doesn't number screens, so you'd be checking the position coordinates of each screen (on each shortcut execution) to determine which number each one corresponds to.
- In order to assign the cmd + ctrl + number keyboard shortcuts, you could use MASShortcut (currently used in the UI), or add in just the code that registers the shortcuts in macOS. For the configuration via Terminal command, I think it would be best to read an integer parameter which would correspond to the configured modifier flag combination. If it's zero, then it's off. We then use that modifier flag combination when registering the keyboard shortcuts.

Let me know if you have any questions.

**Comment by carlos-sarmiento** [comment]
@rxhanson Could you give me code pointers to where these two features are implemented so I can see some code and write my implementation?

**Comment by rxhanson** [comment]
Sure, here's a few things that should get you started. 

- I would start by creating a `WindowCalculation`, called something like `SpecificDisplayCalculation` where your resulting window rectangle on the proper screen would be calculated. You can take a look at `NextPrefDisplayCalculation` as a reference. The `UsableScreens` object passed into the calculate function is what you'll use to determine which screen coordinates you'll return in your calculation.
- For registering the shortcuts and callbacks, it's the least amount of code to just use MASShortcut. Here's the documentation: http://cocoadocs.org/docsets/MASShortcut/2.3.6/. You'll probably want a separate class/struct with static functions for registering/unregistering the keyboard shortcuts specific to this action. 
- UserDefaults for the modifier key flags storage are accessed in Rectangle in the Defaults class. Should be straightforward usage there.

**Comment by 211217613** [comment]
is this similar to moving an app to a certain desktop or space? I'm thinking something like https://help.gnome.org/users/gnome-help/stable/shell-workspaces-switch.html.en

Like #137 ?

If so I would be willing to drop a couple bucks into a fund for this feature.

**Comment by rxhanson** [comment]
@211217613 It's not, although this is commonly requested in a number of closed issues.

> Apple never released a public API for Spaces, so any direct interaction with Spaces uses private APIs that are actually a bit shaky. Using the private API adds enough complexity to the app to where I feel it's better off without it. If Apple decides to release a public API for it, I'll add it in.

**Comment by McoreD** [comment]
In Windows, you can do WinKey+Shift+Left or WinKey+Shift+Right to move a window to the left monitor or right monitor. I am guessing, this is the behavior we refer that being impossible in macOS due to lack of APIs?

**Comment by rxhanson** [comment]
@McoreD Next/previous monitor are already in Rectangle as "Next Display" and "Previous Display". 

It is possible to add support for Spaces (also referred to as Desktops or virtual desktops), but only with private macOS APIs. The private APIs don't necessarily work consistently for all macs, making Spaces a pain to support. That's the main reason why I choose not to include that functionality in Rectangle.

**Comment by McoreD** [comment]
What an amazing implementation! Not only it moves to next monitor, it also maximises correctly in one go. I’ve mapped it to same WinKey + Shift + Left and Right arrow keys and it works perfectly. Thank you!

**Comment by rxhanson** [comment]
@McoreD Thanks for checking it out. I have a todo task for preserving window size ratios when moving to across monitors that I'll hopefully get some time to look at in an upcoming release.

**Comment by McoreD** [comment]
@rxhanson  Apple tried to preserve windows size, and they implemented it poorly.  When you chose windows menu and click to move to other monitor the window goes beyond the monitor for me, that is because Apple didn’t take into account the DPI of that monitor. I’m perfectly happy with the way Rectangle is doing it at the moment which is exactly what I initially wanted: Move to next/previous display and maximise.

**Comment by rafa-suagu** [comment]
> In Windows, you can do WinKey+Shift+Left or WinKey+Shift+Right to move a window to the left monitor or right monitor. I am guessing, this is the behavior we refer that being impossible in macOS due to lack of APIs?

I've used [Swiss Arrows](https://apps.apple.com/es/app/swiss-arrows/id498502402?mt=12) app (dead) from 2012 and has this feature working properly. With this app, you can send the windows to the top/left/right/bottom monitor using the arrows keys.

Since this app is dead, I want to replace it with Rectangle but I love this feature because gives me more flexibility about monitor management. 

@rxhanson do you think that can be achieved? 

I don't have experience with Swift but I can try it to improve your amazing app!

**Comment by rxhanson** [comment]
@rafa-as Sounds like you're looking for some really similar behavior to the original issue here, except with arrow keys instead of numbers. In terms of adding these types of behavior to Rectangle, there's nothing technically preventing them, it could just end up really crowded to have additional shortcuts for all of these behaviors. At a start, I'd definitely consider a pull request that adds this behavior without UI configuration (maybe just with a Terminal command toggle and hard coded shortcut values). If it's a popular enough item, then I would consider adding UI config.

**Comment by McoreD** [comment]
> I've used [Swiss Arrows](https://apps.apple.com/es/app/swiss-arrows/id498502402?mt=12) app (dead) from 2012 and has this feature working properly. With this app, you can send the windows to the top/left/right/bottom monitor using the arrows keys.

Isn't what you are after, here:
![](https://mike.getsharex.com/2020/12/20201208_XCYFhbh4Ru.png)

**Comment by rafa-suagu** [comment]
> > I've used [Swiss Arrows](https://apps.apple.com/es/app/swiss-arrows/id498502402?mt=12) app (dead) from 2012 and has this feature working properly. With this app, you can send the windows to the top/left/right/bottom monitor using the arrows keys.
> 
> Isn't what you are after, here:
> ![](https://camo.githubusercontent.com/7c449f8dbeb625f51da02206486a5282e7f31b954e39a17d0dca1bebd2ffcc5d/68747470733a2f2f6d696b652e6765747368617265782e636f6d2f323032302f31322f32303230313230385f584359466862683452752e706e67)

I'm talking about move windows between top/bottom/left/right monitors, not inside the same monitor.

**Comment by McoreD** [comment]
@rafa-as I am still lost, and I am sorry that I don't understand this request. To send the window to the top/left/right/bottom monitor using the arrow keys, you can customise the hotkeys for these commands:
![](https://mike.getsharex.com/2020/12/20201209_Nep2SzaBJc.png)
Unless you mean, send a window to a specific monitor so that you don't have to cycle through the next/previous monitors?

**Comment by rafa-suagu** [comment]
> Unless you mean, send a window to a specific monitor so that you don't have to cycle through the next/previous monitors?

Exactly @McoreD !

Send the windows to the top/left/right/bottom monitor using the arrow keys.

**Comment by kuncevic** [comment]
Are there any blockers to make this happen?

**Comment by rxhanson** [comment]
Just an FYI that this functionality is now built into the custom shortcut builder of my paid superset of Rectangle: [Hookshot](https://hookshot.app).  

The reason I added it into Hookshot and not Rectangle is merely because it made sense to add it to Hookshot's custom shortcut builder - a feature that I have no plans on porting to Rectangle. Adding this into Rectangle's meant-to-be-simple UI is just not desirable for me, but I would consider a PR for adding something like this into Rectangle if someone came up with an elegant way to do it or built it entirely without the need for UI changes.
